### PR TITLE
ops: revise kokkos matrix-vector product

### PR DESCRIPTION
**Pull request by mzuzek**
refs #446

### CODE REVIEW

File: include/pressio/type_traits/have_matching_exe_space.hpp [hunk 1]
```diff
@@ -49,24 +49,36 @@
...
+  > = std::is_same<
+    typename ::pressio::impl::execution_space<T1>::type,
+    typename ::pressio::impl::execution_space<T2>::type
+    >::value;
+
+} // namespace impl
 
 template <typename ... Args>
-struct have_matching_execution_space;
+struct have_matching_execution_space : public std::false_type {};
```
Comment (fnrizzi) [inline include/pressio/type_traits/have_matching_exe_space.hpp:73]: this seems weird to me, how can this not conflict with the one below T1, T2, ...rest? 
Also, I don't think this ever get instantiated. I would remove the : public std::false_type{}. Not needed.

File: include/pressio/type_traits/traits_tpl.hpp [hunk 2]
```diff
@@ -77,6 +77,17 @@ template <typename T, typename Enabled=void> struct DeviceType;
...
 template <typename T, typename Enabled=void> struct execution_space;
 
+template<typename T, typename Enable=void>
+constexpr bool has_execution_space_v = false;
+
+template<typename T>
+constexpr bool has_execution_space_v<
+  T,
+  ::pressio::mpl::void_t<
+    decltype(sizeof(::pressio::impl::execution_space<T>))
```
Comment (fnrizzi) [inline include/pressio/type_traits/traits_tpl.hpp:87]: i don't really understand what is this?

File: include/pressio/expressions/impl/diag_traits.hpp [hunk 3]
```diff
@@ -110,9 +110,14 @@ struct DiagTraits<
...
-struct execution_space<::pressio::expressions::impl::DiagExpr<T>>
+struct execution_space<
+  ::pressio::expressions::impl::DiagExpr<T>,
+  ::pressio::mpl::enable_if_t<
+    ::pressio::impl::has_execution_space_v<T>
+    >
+  >
 {
-  using type = typename T::traits::execution_space;
+  using type = typename execution_space<T>::type;
```
Comment (fnrizzi) [inline include/pressio/expressions/impl/diag_traits.hpp:120]: what is going on here?

**Comment by fnrizzi** [comment]
I find the use of `impl::execution_space` and `has_execution_space` confusing.

**Comment by mzuzek** [comment]
@fnrizzi As we have discussed and planned, I rebased this PR on #502 which removes `have_matching_execution_space` predicate all together. This obsoletes c40b2c473698dc87184d7eaccfcd8b36ae0a62cf (making it empty) and hopefully resolves all raised issues and questions on  `execution_space` and `has_execution_space` utilities which supported that predicate.

**Comment by mzuzek** [comment]
for the record, we had weird [CI failure](https://github.com/Pressio/pressio/actions/runs/3535815865/jobs/5934242375) that did not reproduce on the re-run:
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  stoi
```

File: include/pressio/ops/kokkos/ops_level2.hpp [hunk 4]
```diff
@@ -63,65 +63,129 @@ namespace pressio{ namespace ops{
...
 	const A_type & A,
 	const x_type & x,
-	const scalar_type beta,
+	const beta_t beta,
 	y_type & y)
 {
   assert( y.extent(0) == A.extent(0) );
   assert( A.extent(1) == x.extent(0) );
+
+  // Kokkos requires alpha to be of A scalar type and beta to be of y scalar type
```
Comment (fnrizzi) [inline include/pressio/ops/kokkos/ops_level2.hpp:99]: does this really matter here since they are all converted to the same scalar type?

Comment (mzuzek) [reply include/pressio/ops/kokkos/ops_level2.hpp]: right, please check out scalar type simplification on b7173dbf981a86efb411e27a9ffc4521c9a73395 (based on that it's the same)

File: include/pressio/ops/kokkos/ops_level2.hpp [hunk 5]
```diff
@@ -63,65 +63,123 @@ namespace pressio{ namespace ops{
...
 	const A_type & A,
 	const x_type & x,
-	const scalar_type beta,
+	const beta_t beta,
 	y_type & y)
 {
   assert( y.extent(0) == A.extent(0) );
   assert( A.extent(1) == x.extent(0) );
+
+  using sc_t = typename A_type::non_const_value_type;
```
Comment (fnrizzi) [inline include/pressio/ops/kokkos/ops_level2.hpp:99]: this won`t work if A_type is a pressio expression

File: include/pressio/ops/kokkos/ops_level2.hpp [hunk 6]
```diff
@@ -63,65 +63,123 @@ namespace pressio{ namespace ops{
...
 	const A_type & A,
 	const x_type & x,
-	const scalar_type beta,
+	const beta_t beta,
 	y_type & y)
 {
   assert( y.extent(0) == A.extent(1) );
   assert( A.extent(0) == x.extent(0) );
+
+  using sc_t = typename A_type::non_const_value_type;
```
Comment (fnrizzi) [inline include/pressio/ops/kokkos/ops_level2.hpp:160]: same as above

File: include/pressio/ops/kokkos/ops_level2.hpp (hunk 5 above)
Comment (fnrizzi) [reply include/pressio/ops/kokkos/ops_level2.hpp:99]: i am surprised the tests even compile.... unless i miss something evident

File: include/pressio/ops/kokkos/ops_level2.hpp [hunk 7]
```diff
@@ -63,65 +63,123 @@ namespace pressio{ namespace ops{
...
   (::pressio::is_native_container_kokkos<x_type>::value or
    ::pressio::is_expression_acting_on_kokkos<x_type>::value) and
   (::pressio::is_native_container_kokkos<y_type>::value or
     ::pressio::is_expression_acting_on_kokkos<y_type>::value) and
-  ::pressio::Traits<A_type>::rank == 2 and
-  ::pressio::Traits<x_type>::rank == 1 and
-  ::pressio::Traits<y_type>::rank == 1
+  // scalar compatibility
+  ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value and
+  std::is_convertible<alpha_t, typename A_type::non_const_value_type>::value and
```
Comment (fnrizzi) [inline include/pressio/ops/kokkos/ops_level2.hpp:84]: see comment below, i am surprised this overload gets selected since it should compile

Comment (fnrizzi) [reply include/pressio/ops/kokkos/ops_level2.hpp:84]: also why you are not using the Traits::scalar_type like you did for Eigen?

Comment (mzuzek) [reply include/pressio/ops/kokkos/ops_level2.hpp]: good point, thanks for catching this: fixed with 1fb60febaea4bbed9207bc3e793fc98fc8d24e6b

File: include/pressio/ops/kokkos/ops_level2.hpp (hunk 5 above)
Comment (mzuzek) [reply include/pressio/ops/kokkos/ops_level2.hpp]: we do test `pressio::span` for `x` but not for `y` (that wouldn't have compiled)

**Comment by mzuzek** [comment]
@fnrizzi 
### Kokkos test coverage

Both `nontranspose` and `transpose` overloads of `pressio::ops::product()` are tested with all combinations of plain views and available Pressio expressions:
| `A` | `x` | `y` |
|:---:|:---:|:---:|
| view2d | view1d | view1d |
| view2d | _span_ | view1d |
| view2d | _diag_ | view1d |
| view2d | view1d | _span_ |
| view2d | view1d | _diag_ |
| view2d | _span_ | _span_ |
| view2d | _span_ | _diag_ |
| view2d | _diag_ | _span_ |
| view2d | _diag_ | _diag_ |
| _subspan_ | view1d | view1d |
| _subspan_ | _span_ | view1d |
| _subspan_ | _diag_ | view1d |
| _subspan_ | view1d | _span_ |
| _subspan_ | view1d | _diag_ |
| _subspan_ | _span_ | _span_ |
| _subspan_ | _span_ | _diag_ |
| _subspan_ | _diag_ | _span_ |
| _subspan_ | _diag_ | _diag_ |

> **Note:** `A` could be also `pressio::as_diagonal_matrix()` in future - once it's implemented, see #304

File: include/pressio/ops/kokkos/ops_level2.hpp [hunk 8]
```diff
@@ -63,65 +63,123 @@ namespace pressio{ namespace ops{
...
    ::pressio::is_expression_acting_on_kokkos<x_type>::value) and
   (::pressio::is_native_container_kokkos<y_type>::value or
     ::pressio::is_expression_acting_on_kokkos<y_type>::value) and
-  ::pressio::Traits<A_type>::rank == 2 and
-  ::pressio::Traits<x_type>::rank == 1 and
-  ::pressio::Traits<y_type>::rank == 1
+  // scalar compatibility
+  ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value and
+  std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value and
+  std::is_convertible<beta_t, typename ::pressio::Traits<y_type>::scalar_type>::value and
```
Comment (fnrizzi) [inline include/pressio/ops/kokkos/ops_level2.hpp:85]: to be picky, techically this should be also <A_type>::scalar_type

File: include/pressio/ops/kokkos/ops_level2.hpp [hunk 9]
```diff
@@ -63,65 +63,123 @@ namespace pressio{ namespace ops{
...
```
Comment (fnrizzi) [inline include/pressio/ops/kokkos/ops_level2.hpp:146]: same comment as above

**Comment by fnrizzi** [comment]
@mzuzek 
- the table is awesome !
- i only found two things, if you can please fix those i will then merge it

File: include/pressio/ops/kokkos/ops_level2.hpp (hunk 8 above)
Comment (mzuzek) [reply include/pressio/ops/kokkos/ops_level2.hpp]: thanks, fixed with 77d2b022777039fe56ddcfc6e03d778296e7ca4d

**Review by fnrizzi — APPROVED** [review]

**Pull request merged**